service/controller/clusterapi/v29/resource/tccp/template_render_test.go

```diff
@@ -0,0 +1,273 @@
+package tccp
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"unicode"
+
+	g8sv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/cluster/v1alpha1"
+	"github.com/giantswarm/micrologger/microloggertest"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/giantswarm/aws-operator/pkg/annotation"
+	"github.com/giantswarm/aws-operator/pkg/label"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/detection"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/key"
+)
+
+const (
+	TestClusterID           = "8y5ck"
+	TestMachineDeploymentID = "al9qy"
+)
+
+var update = flag.Bool("update", false, "update .golden CF template file")
```

Comment:
I like the idea but how it is implemented is wrong IMO. Flags should only be handled in the program start and having particular flags for certain tests increases complexity immensely. It is super hard for me personally to remember such exceptions. I usually just to `go test ./...` and consider this being fine. 

As I already mentioned in our conversation, I think when there is such a functionality, which is legit, it should be more properly incorporated into the command line structure like this. 

```
aws-operator daemon
aws-operator template
aws-operator version
```

We could have a new subcommand for all this magic and have a properly defined way of generating templates and do whatever we want. Mixing this with unit tests that way is something I never saw before and it does not feel right to me. This here is now just magically hidden somewhere and you really have to know about it. This is designed to break that way when you ask me.

Reply:
I strongly disagree here. Few reasons:

- Template rendering function here is private function. Only rendering template without all other wiring would force it to be public function which would be quite wrong from the normal design and execution flow point of view. If it was kept as private function, then it requires all the wiring and configuration and execution of that `aws-operator template` would either require a lot of runtime environment configuration which makes it very cumbersome and defeats the whole purpose, or alternatively it introduces quite a lot of code in completely unrelated part of `aws-operator` that must be maintained. I don't like any of this.

- Concept of `golden test files` is widely accepted / known these days, as I pointed out above so I don't see anything wrong with the concept per se.

- The flag here doesn't add complexity at all. You don't need to know about it at all. If you break the test, you obviously go to that test and see what broke. There is a comment that directly explains what you need to do in order to bring it up to date.
  - Note: If you run `go test ./... -update` - all tests besides this will actually give you a warning about unregistered flag and that's something you should not do.

- Your approach doesn't prevent accidental changes toward template rendering nor does it provide easy solution to add different test cases for different template rendering conditions (different combinations of availability zones, potential effects of different number of machine deployments for the cluster).

Reply:
I personally like that. I don't see what is supposed to break here. This is a test. I would just ad note in the README that you can update golden files with `go test package-here -update` otherwise this kind of hidden functionality easy to forget.

Reply:
I don't like having flag on test and having a test which modify a local file feels wrong to me, but having read about golden file I see where this is coming from, so I'm not standing in your way here.

Reply:
I suggest to make the rendering a first class citizen. Private functions can be made public. The golden test file can be updated using the new subcommand. We do not need a weird flag in unit tests that nobody knows about. Saying to add complexity does not increase complexity is invalid. Unit tests can simply make use of template rendering wherever needed. I do not question the concept, I question the implementation, which is not the best we can do here. Quite the contrary.

Reply:
> I suggest to make the rendering a first class citizen. Private functions can be made public.

I can only imagine the weeks this would take work.

> The golden test file can be updated using the new subcommand. We do not need a weird flag in unit tests that nobody knows about.

For me this is several folds more complex. The same applies for every newcomer, because this is something that is not idiomatic according to general way to do it. I don't also see any benefits in this approach. Only downsides.

Reply:
I give up now. I made my points. Disagree and commit.